Coord print

pp-mo opened the pull request:
Closes #4488, #4494
Linked to https://github.com/SciTools/iris/discussions/4501

This is just an initial draft version, for discussing the new behaviour + printed output format

**Please review for behaviours (output results), and not the code (yet!)**


Some general notes :
  * I tried to mimic the general form of the `Cube.__str__` output
  * the `__repr__` produces a minimal one-line form, and the `__str__` a detailed multiline printout, like the Cube ones

pp-mo commented:
Sample results :
```
$ python simple_coord_print_tests.py 

Small section of latitude coord[:2]
------------REPR:-------
<DimCoord: latitude / (degrees)  [90. , 87.5]>
------------STR:--------
DimCoord :  latitude / (degrees)
    [89.999985, 87.499985]
    shape: (2,)
    dtype: float32
    standard_name: 'latitude'
------------------------

co[:2] with 4x bounds + printoptions(precision=3) 
------------REPR:-------
<DimCoord: latitude / (degrees)  [90. , 87.5]>
------------STR:--------
DimCoord :  latitude / (degrees)
    [90. , 87.5]
    bounds=[
        [90.  , 88.75, 90.  , 88.75],
        [88.75, 86.25, 88.75, 86.25]]
    shape: (2,) ; bounds=(2, 4)
    dtype: float32
    standard_name: 'latitude'
------------------------


Longer section co[:5]
------------REPR:-------
<DimCoord: latitude / (degrees)  [90. , 87.5, 85. , 82.5, 80. ]>
------------STR:--------
DimCoord :  latitude / (degrees)
    [89.999985, 87.499985, 84.999985, 82.499985, 79.99999 ]
    shape: (5,)
    dtype: float32
    standard_name: 'latitude'
------------------------


Longer section co[:5] with bounds
------------REPR:-------
<DimCoord: latitude / (degrees)  [90. , 87.5, 85. , 82.5, 80. ]>
------------STR:--------
DimCoord :  latitude / (degrees)
    [89.999985, 87.499985, 84.999985, 82.499985, 79.99999 ]
    bounds=[
        [90.      , 88.749985],
        [88.749985, 86.249985],
        [86.249985, 83.749985],
        [83.749985, 81.249985],
        [81.249985, 78.74999 ]]
    shape: (5,) ; bounds=(5, 2)
    dtype: float32
    standard_name: 'latitude'
------------------------


Still longer co[-15:] with bounds
------------REPR:-------
<DimCoord: latitude / (degrees)  [-55. , -57.5, ..., -87.5, -90. ], shape=(15,)>
------------STR:--------
DimCoord :  latitude / (degrees)
    [-54.99994, -57.49994, -59.99994, -62.49994, -64.99994,
     -67.49994, -69.99994, -72.49994, -74.99994, -77.49994,
     -79.99994, -82.49994, -84.99994, -87.49992, -89.99992]
    bounds=[
        [-53.749947, -56.24994 ],
        [-56.24994 , -58.74994 ],
        ...,
        [-86.24993 , -88.74992 ],
        [-88.74992 , -90.      ]]
    shape: (15,) ; bounds=(15, 2)
    dtype: float32
    standard_name: 'latitude'
------------------------


Full co
------------REPR:-------
<DimCoord: latitude / (degrees)  [ 90. , 87.5, ..., -87.5, -90. ], shape=(73,)>
------------STR:--------
DimCoord :  latitude / (degrees)
    [ 89.999985,  87.499985, ..., -87.49992 , -89.99992 ]
    shape: (73,)
    dtype: float32
    standard_name: 'latitude'
------------------------


Full co with bounds
------------REPR:-------
<DimCoord: latitude / (degrees)  [ 90. , 87.5, ..., -87.5, -90. ], shape=(73,)>
------------STR:--------
DimCoord :  latitude / (degrees)
    [ 89.999985,  87.499985, ..., -87.49992 , -89.99992 ]
    bounds=[
        [ 90.      ,  88.749985],
        [ 88.749985,  86.249985],
        ...,
        [-86.24993 , -88.74992 ],
        [-88.74992 , -90.      ]]
    shape: (73,) ; bounds=(73, 2)
    dtype: float32
    standard_name: 'latitude'
------------------------


Small co with lazy points
dask.array<array, shape=(4,), dtype=float32, chunksize=(4,), chunktype=numpy.ndarray>
------------REPR:-------
<AuxCoord: latitude / (degrees)  <lazy>, shape=(4,)>
------------STR:--------
AuxCoord :  latitude / (degrees)
    <lazy>
    shape: (4,)
    dtype: float32
    standard_name: 'latitude'
------------------------


Small co with real points but lazy bounds
------------REPR:-------
<AuxCoord: latitude / (degrees)  <lazy>, shape=(4,)>
------------STR:--------
AuxCoord :  latitude / (degrees)
    <lazy>
    shape: (4,) ; bounds=(4, 2)
    dtype: float32
    standard_name: 'latitude'
------------------------


Float with masked points
------------REPR:-------
<AuxCoord: latitude / (degrees)  [89.999985, -- , ..., 69.99999 , 67.5 ], shape=(10,)>
------------STR:--------
AuxCoord :  latitude / (degrees)
    [89.999985, --       , 84.999985, --       , 79.99999 ,
     77.49999 , 74.99999 , 72.49999 , 69.99999 , 67.5     ]
    shape: (10,)
    dtype: float32
    standard_name: 'latitude'
------------------------


Integer
------------REPR:-------
<AuxCoord: integer_points / (unknown)  [0, 1, 2, 3, 4]>
------------STR:--------
AuxCoord :  integer_points / (unknown)
    [0, 1, 2, 3, 4]
    shape: (5,)
    dtype: int64
    long_name: 'integer_points'
------------------------


Integer with masked points
------------REPR:-------
<AuxCoord: integer_points / (unknown)  [0 , --, 2 , --, 4 ]>
------------STR:--------
AuxCoord :  integer_points / (unknown)
    [0 , --, 2 , --, 4 ]
    shape: (5,)
    dtype: int64
    long_name: 'integer_points'
------------------------


Longer integers with masked points
------------REPR:-------
<AuxCoord: integer_points / (unknown)  [995 , -- , ..., 1018, 1019], shape=(25,)>
------------STR:--------
AuxCoord :  integer_points / (unknown)
    [995 , --  , ..., 1018, 1019]
    shape: (25,)
    dtype: int64
    long_name: 'integer_points'
    var_name: 'qq'
------------------------


dates with bounds
------------REPR:-------
<AuxCoord: x / (days since 2015-05-17)  shape=(10,)>
------------STR:--------
AuxCoord :  x / (days since 2015-05-17)
    [2015-05-17 00:00:00, 2015-05-28 02:40:00,
     2015-06-08 05:20:00, 2015-06-19 08:00:00,
     2015-06-30 10:40:00, 2015-07-11 13:20:00,
     2015-07-22 16:00:00, 2015-08-02 18:40:00,
     2015-08-13 21:20:00, 2015-08-25 00:00:00]
    bounds=[
        [2015-05-11 10:40:00, 2015-05-22 13:20:00],
        [2015-05-22 13:20:00, 2015-06-02 16:00:00],
        ...,
        [2015-08-08 08:00:00, 2015-08-19 10:40:00],
        [2015-08-19 10:40:00, 2015-08-30 13:20:00]]
    shape: (10,) ; bounds=(10, 2)
    dtype: float64
    var_name: 'x'
    attributes: {'a': 14, 'b': None}
------------------------


dates with masked points
------------REPR:-------
<AuxCoord: x / (days since 2015-05-17)  shape=(10,)>
------------STR:--------
AuxCoord :  x / (days since 2015-05-17)
    [2015-05-17 00:00:00, 2015-05-28 02:40:00,
     --                 , 2015-06-19 08:00:00,
     2015-06-30 10:40:00, --                 ,
     2015-07-22 16:00:00, 2015-08-02 18:40:00,
     2015-08-13 21:20:00, 2015-08-25 00:00:00]
    bounds=[
        [2015-05-11 10:40:00, 2015-05-22 13:20:00],
        [2015-05-22 13:20:00, 2015-06-02 16:00:00],
        ...,
        [2015-08-08 08:00:00, 2015-08-19 10:40:00],
        [2015-08-19 10:40:00, 2015-08-30 13:20:00]]
    shape: (10,) ; bounds=(10, 2)
    dtype: float64
    var_name: 'x'
    attributes: {'a': 14, 'b': None}
------------------------


multidimensional
------------REPR:-------
<AuxCoord: unknown / (unknown)  shape=(7, 5, 4)>
------------STR:--------
AuxCoord :  unknown / (unknown)
    [[[2.06187124e+00, 2.62729909e+02, 5.04530999e-01,
       7.72296297e-02],
      [1.98149735e+02, 2.72691428e+00, 5.59201397e-01,
       9.15158572e-02],
      ...,
      [9.35959507e-02, 9.87648075e+01, 1.41555682e+02,
       1.19528981e+01],
      [3.96118914e+00, 5.55304391e-02, 3.79748850e+02,
       1.38359365e+01]],
    
     [[2.88366966e+00, 3.93076950e+01, 5.14994973e+00,
       2.43560359e+00],
      [2.29709442e+00, 5.93655905e-01, 1.12286999e-01,
       1.59593353e-01],
      ...,
      [3.64811663e+02, 2.43749085e+00, 5.07759208e-01,
       5.24134052e+01],
      [1.93544347e+00, 4.15162702e+00, 3.19379729e+02,
       2.75618918e+00]],
    
     ...,
    
     [[1.31830607e+00, 2.51751930e+00, 3.66400899e-01,
       1.76748038e+00],
      [3.85791606e+00, 1.43061739e-01, 2.38357293e-01,
       5.11634593e+00],
      ...,
      [1.43894872e+02, 2.35019027e+01, 5.90061070e+01,
       1.88762605e+02],
      [3.44989074e+01, 1.06243423e+00, 1.14218976e+00,
       7.91787038e+01]],
    
     [[9.12835821e+00, 2.49120964e+02, 1.15961382e+00,
       1.05658599e+02],
      [1.99595152e-01, 3.33424239e+01, 3.64210805e+02,
       3.50297194e+01],
      ...,
      [1.26869326e+02, 3.58661096e+00, 1.92530940e+00,
       5.61565460e+01],
      [1.00202310e+00, 6.23401913e+00, 1.36128130e+01,
       6.95941520e+00]]]
    shape: (7, 5, 4)
    dtype: float64
------------------------


small multidim
------------REPR:-------
<AuxCoord: unknown / (unknown)  [[ 2.062, 262.73 ], [198.15 , 2.727]]>
------------STR:--------
AuxCoord :  unknown / (unknown)
    [[  2.06187124, 262.72990877],
     [198.14973544,   2.72691428]]
    shape: (2, 2)
    dtype: float64
------------------------


SCALAR:
------------REPR:-------
<AuxCoord: forecast_period / (hours)  [15.]>
------------STR:--------
AuxCoord :  forecast_period / (hours)
    [15.]
    shape: (1,)
    dtype: float64
    standard_name: 'forecast_period'
------------------------


==== REAL DATA nc test cube coords ...
------------REPR:-------
<DimCoord: time / (days since 1850-01-01)  shape=(4,)>
------------STR:--------
DimCoord :  time / (days since 1850-01-01)
    [2046-01-01 12:00:00, 2046-01-02 12:00:00,
     2046-01-03 12:00:00, 2046-01-04 12:00:00]
    bounds=[
        [2046-01-01 00:00:00, 2046-01-02 00:00:00],
        [2046-01-02 00:00:00, 2046-01-03 00:00:00],
        [2046-01-03 00:00:00, 2046-01-04 00:00:00],
        [2046-01-04 00:00:00, 2046-01-05 00:00:00]]
    shape: (4,) ; bounds=(4, 2)
    dtype: float64
    standard_name: 'time'
    long_name: 'time'
    var_name: 'time'
------------------------

------------REPR:-------
<DimCoord: latitude / (degrees)  [-89.142, -88.029, ..., 88.029, 89.142], shape=(160,)>
------------STR:--------
DimCoord :  latitude / (degrees)
    [-89.14151943, -88.02942887, ...,  88.02942887,
      89.14151943]
    bounds=[
        [-90.        , -88.62472856],
        [-88.62472856, -87.49119059],
        ...,
        [ 87.49119059,  88.62472856],
        [ 88.62472856,  90.        ]]
    shape: (160,) ; bounds=(160, 2)
    dtype: float64
    standard_name: 'latitude'
    long_name: 'latitude'
    var_name: 'lat'
------------------------

------------REPR:-------
<DimCoord: longitude / (degrees)  [ 0. , 1.125, ..., 357.75 , 358.875], shape=(320,)>
------------STR:--------
DimCoord :  longitude / (degrees)
    [  0.   ,   1.125, ..., 357.75 , 358.875]
    bounds=[
        [ -0.5625,   0.5625],
        [  0.5625,   1.6875],
        ...,
        [357.1875, 358.3125],
        [358.3125, 359.4375]]
    shape: (320,) ; bounds=(320, 2)
    dtype: float64
    standard_name: 'longitude'
    long_name: 'longitude'
    var_name: 'lon'
------------------------

------------REPR:-------
<DimCoord: height / (m)  [10.]>
------------STR:--------
DimCoord :  height / (m)
    [10.]
    shape: (1,)
    dtype: float64
    standard_name: 'height'
    long_name: 'height'
    var_name: 'height'
    attributes: {'positive': 'up'}
------------------------

```

rcomer commented:
I just tried this out with a file I had lying around.  I think it's a big improvement as there is less mess all over the screen!  I also really like that you include the shapes and dtype.

Sometimes, though it is useful to print the lot out in order to trawl through and figure out what's going wrong!  Of course we can print the points and bounds separately and for time coordinates I was thinking I could do
```python
print(coord.units.num2pydate(coord.points))
```
but that gave me this, which I think is slightly hideous:
```
[cftime.DatetimeGregorian(1993, 10, 25, 0, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1993, 11, 1, 0, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1993, 11, 9, 0, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1994, 10, 25, 0, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1994, 11, 1, 0, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1994, 11, 9, 0, 0, 0, 0, has_year_zero=False)
...etc...
```

Is there a way the user could easily access the time-formatting used by the coord `__str__` and apply it to the arrays?

rcomer commented:
Never mind.  I just realised I can do

```python
print(coord.units.num2date(coord.points).astype(str))
```

pp-mo commented:
> Is there a way the user could easily access the time-formatting used by the coord __str__

It's an interesting point, and I think there's still scope for a helper to do this.
However it needs the context of the Coord units, yet for consistency should operate onn the specific coord  points/bounds.
So it would probably be something like `Coord.points_dates` / `Coord.bounds_dates`.
I'm not sure if I think that is a great idea or not.

Another way would be to  provide a utility equivalent to the `num2date(...).astype(str)` you did above, E.G. points_to_dates(array, units).  But maybe the oneliner is OK, and at most needs documenting as  handy wrinkle ?

rcomer commented:
> Another way would be to provide a utility equivalent to the `num2date(...).astype(str)` you did above, E.G. points_to_dates(array, units). But maybe the oneliner is OK, and at most needs documenting as handy wrinkle ?

Would that be a job for `cf_units` rather than `iris`?  E.g. `Unit.num2datestr(array)`, though it would only save the user from typing `astype(str)` so I'm not sure whether it's worth it.  This also reminds me of #4177 where we decided that such conveniences were overdoing it....

pp-mo commented:
Mods trying out some newer ideas -- see https://github.com/SciTools/iris/discussions/4501#discussioncomment-1962816

rcomer commented:
> 
>  something like `Coord.points_as_dates()` / `Coord.bounds_as_dates()`. I'm not sure if I think that is a great idea or not.
> 
Having slept on it, how about something like `Coord.print_values(array)`.  Then I can pass it `Coord.points` or `Coord.bounds` and it should be transparent to me that I am realising any lazy arrays.

pp-mo commented:
some new tweaks.
Still not ready : needs proper tests, _and_ some consideration (at least initial) for  printout of _other_ _DimMeta subclasses 
I think that is : CellMeasure / AncillaryVariable / MeshCoord

pp-mo commented:
> @rcomer how about something like Coord.print_values(array)? 

Still thinking about that, but let's make it a separate issue ?

pp-mo commented:
In case anyone doesn't get a ping from the Discussion, I just raised there an awkward question of column formatting ...
https://github.com/SciTools/iris/discussions/4501#discussioncomment-1985700

pp-mo commented:
( Note: rebased onto upstream/main ; removed some test-debug code )

Still TODO:
  * full testing of more aspects
    * size-related partial-array printout types
    * alternative cases : lazy, masked
    * non-Coord operations : CellMeasure; AncillaryVariable; MeshCoord
  * reorganise attribute printouts, in line with the Cube version (cf [comment](https://github.com/SciTools/iris/discussions/4501#discussioncomment-1986399))
  * fix MeshCoord and Mesh printout (maybe in another PR ?)

pp-mo commented:
Finally done !
There was a lot of knock-on here : as well as fixing various tests, I couldn't really defer the issue of Mesh printout, since I couldn't get this passing without doing something about that.

There is one possibly outstanding issue, regarding the printout of attributes... (separate comment follows)

pp-mo commented:
Currently I am just printing attributes as a repr, and hoping it fits on a line
 -- true for both _DimensionalMetadata things (mostly Coords), and the Mesh

@lbdreyer already suggested that this might be nasty in some cases (like long comment or history-type attributes)
In which case, output more like that of cube attributes would be better.
I think we can address that separately : I have some prototype code for it,  and I will submit something if there is time.

lbdreyer reviewed (commented):
I am still looking through the last bits of the mesh parts of the changes. In case I don't manage to finish it all today, I at least wanted to submit what I had. 

Looking very good so far! Just a few minor comments I noticed while going through it

pp-mo commented on lib/iris/coords.py:
```diff
@@ -268,78 +272,247 @@ def _has_lazy_values(self):
+        max_array_width=None,
+    ):
+        """
+        Make a printable text summary of a dimensional cube component.
```
Not sure of creating the string dynamically.  It _might_ work in Sphinx, or not ?
I will have a try ...

pp-mo replied:
No, I think it's not possible to produce a different docstring for each subclass without a metaclass type of operation (post-definition), which I think is too much fuss.

I see we do have some operations which describe themselves as 'dimensional metadata'.  E.G. file:///net/home/h05/itpp/git/iris/iris_main/docs/src/_build/html/generated/api/iris/coords.html?highlight=dimensional%20metadata#iris.coords.CellMeasure.copy

But I think just leaving it un-named is easier, as you suggest.

pp-mo commented on lib/iris/coords.py:
```diff
@@ -268,78 +272,247 @@ def _has_lazy_values(self):
+        # Implement conditional defaults for control args.
+        if max_values is None:
+            max_values = 5 if shorten else 15
+        precision = 3 if shorten else None
```
I found it was easiest to establish --and explain-- a direct relationship between the controls here and the relevant numpy printoptions.  
https://github.com/SciTools/iris/pull/4499/commits/cd92237c2e2eb2d59a555b05152981d5c8c8c0eb
This is anyway more direct now than in earlier code versions, so I was able to use the numpy names for (most of) the controls.  I baulked at renaming 'max_values' as 'threshold' though, as it seemed such an unhelpful name !

lbdreyer commented on lib/iris/experimental/ugrid/mesh.py:
```diff
@@ -975,77 +925,129 @@ def __ne__(self, other):
+                element_exists = main_conn is not None
+            if element_exists:
+                # Include a section for this element
+                line(element, 1)
```
```suggestion
                line(f"{element}:", 1)
```

pp-mo replied:
TBH I'm not too sure about this.  I considered + rejected this, initially, because I'm concerned about having colons "everywhere".
Since there  are effectively sections + sub-sections here (e.g. "node coordinates"), and the individual coords + connectivities also use colons within the reprs, we would wind up with a whole lot of them.

If we do do this, I think we should be consistent and add it into the other 'subsection' headings.
BTW the [re-format of the attributes listing](https://github.com/SciTools/iris/discussions/4501#discussioncomment-2034628) will affect this as well.  I'm just going to use a best-guess for that here ..

Trying it out, I also think that 'spacing off' the colon makes sense, when the following content is below and not on the same line
For example:
```
Mesh : 'my_topology_mesh'
    topology_dimension: 2
    node :
        node_dimension: 'NodeDim'
        node coordinates :
            <AuxCoord: longitude / (unknown)  [...]  shape(3,)>
            <AuxCoord: latitude / (unknown)  [...]  shape(3,)>
    face :
        face_dimension: 'FaceDim'
        face_node_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(1, 3)>
        face coordinates :
            <AuxCoord: longitude / (unknown)  [...]>
            <AuxCoord: latitude / (unknown)  [...]>
    optional connectivities :
        face_face_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(1, 3)>
        face_edge_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(1, 3)>
        edge_face_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(3, 2)>
    long_name: 'my_topology_mesh'
    var_name: 'mesh'
    attributes :
        notes        'this is a test'
        history      'created in my model'
        version      2.3
        Conventions  'CF 1.9; UGRID 1.0'
```

Well... it's not rubbish, but I'm not sure it helps any.
On the other hand, is is probably more consistent with what we now have in the Cube printout
E.G.
```
unknown / (unknown)                 (-- : 1)
    Auxiliary coordinates:
        my_aux                          x
    Attributes:
        dict                        {'a': 1, 'b': 'two'}
        float                       3.2
        int                         9
        string key                  this info
```

So.. there we currently have a colon (not spaced-off) for each 'section title', 
and no colons between key + value in the attributes
But there are no 'subsections' in that case, which is maybe what makes the above look a bit messy ?

pp-mo replied:
P.S. Now I've finally committed that, I'm not sure about the 'spacing' off' either.
Here's the above without that ..
```
Mesh : 'my_topology_mesh'
    topology_dimension: 2
    node:
        node_dimension: 'NodeDim'
        node coordinates:
            <AuxCoord: longitude / (unknown)  [...]  shape(3,)>
            <AuxCoord: latitude / (unknown)  [...]  shape(3,)>
    face:
        face_dimension: 'FaceDim'
        face_node_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(1, 3)>
        face coordinates:
            <AuxCoord: longitude / (unknown)  [...]>
            <AuxCoord: latitude / (unknown)  [...]>
    optional connectivities:
        face_face_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(1, 3)>
        face_edge_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(1, 3)>
        edge_face_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(3, 2)>
    long_name: 'my_topology_mesh'
    var_name: 'mesh'
    attributes:
        notes        'this is a test'
        history      'created in my model'
        version      2.3
        Conventions  'CF 1.9; UGRID 1.0'
```
.. and here with no section colons at all..
```
Mesh : 'my_topology_mesh'
    topology_dimension: 2
    node
        node_dimension: 'NodeDim'
        node coordinates
            <AuxCoord: longitude / (unknown)  [...]  shape(3,)>
            <AuxCoord: latitude / (unknown)  [...]  shape(3,)>
    face
        face_dimension: 'FaceDim'
        face_node_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(1, 3)>
        face coordinates
            <AuxCoord: longitude / (unknown)  [...]>
            <AuxCoord: latitude / (unknown)  [...]>
    optional connectivities
        face_face_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(1, 3)>
        face_edge_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(1, 3)>
        edge_face_connectivity: <Connectivity: unknown / (unknown)  [...]  shape(3, 2)>
    long_name: 'my_topology_mesh'
    var_name: 'mesh'
    attributes
        notes        'this is a test'
        history      'created in my model'
        version      2.3
        Conventions  'CF 1.9; UGRID 1.0'
```

I'm quite conflicted on this now

lbdreyer replied:
I think I'm coming around to your original format, with no colons. Describing it as sections and subsections helps. I was coming at it from the perspective of what a regular coordinate prints out, but this is slightly different.

I'm happy to stick with how you have it

lbdreyer commented on lib/iris/tests/unit/experimental/ugrid/mesh/test_MeshCoord.py:
```diff
@@ -268,42 +269,86 @@ def setUp(self):
         meshcoord = sample_meshcoord(mesh=self.mesh, location="edge", axis="y")
         result = str(meshcoord)
-        re_expected = r", location='edge', axis='y'"
+        # re_expected = r", location='edge', axis='y'"
```
Were you planning to removed this?

lbdreyer commented on lib/iris/coords.py:
```diff
@@ -268,78 +272,269 @@ def _has_lazy_values(self):
+                        data_str = "[...]"
+
+                if self.has_bounds():
+                    data_str += "+bnds"
```
Can I persuade you to change this to 
```suggestion
                    data_str += "+bounds"
```
It is only two characers after all, and I would prefer we use the official term "bounds" rather than invent a new term "bnds". Dropping the vowels may be natural and obvious to us but I would be concerned that this may trip some people up (non-native english speakers, perhaps?). 

I think it's fine to use `bnds` to name a variable in some example code but not in part of the code in this way

For two vowels I would say it is worth being explicit

pp-mo commented on lib/iris/experimental/ugrid/mesh.py:
```diff
@@ -2943,62 +2945,59 @@ def __eq__(self, other):
+        # Modify the generic 'default-form' result to produce what we want.
+        if shorten:
+            # Single-line form : insert the mesh+location before the array part
+            i_array = result.index("[")
```
Actually, the multiline case uses a search for 'location:', which in principle is also not solid.
Recalling that this was the problem we had with the 'old' `Cube.__str__` output, where `repr_html` was pulling it apart again and searching for certain pieces.
So I have now added some very crude (private) "structural information" output instead, to support the fiddling that `MeshCoord.summary` needs to do.

pp-mo commented on lib/iris/tests/test_coord_api.py:
```diff
@@ -178,7 +178,9 @@ def test_complex(self):
-        self.lat = iris.tests.stock.realistic_4d().coord("grid_latitude")[:10]
+        cube = iris.tests.stock.realistic_4d()
+        self.lat = cube.coord("grid_latitude")[:10]
+        self.height = cube.coord("level_height")[:10]
```
These tests were a bit of a mess : these 4 things were 2 duplicates of only 2 things.
Unfortunately I'm not sure I've improved things much, though.
'height' is not actually an AuxCoord (class), though it is an aux-coord in the cube.

pp-mo commented:
Many thanks @lbdreyer . 
I think I have now addressed all your outstanding points, but I did add a bit of complexity so you may like to consider that.

pp-mo commented:
> I think I have now addressed all your outstanding points, but I did add a bit of complexity so you may like to consider that.

Viewed from the morning after, I think the core routine is less clear now, and I'm sure that can be improved.
On the other hand, we may have more important concerns right now.  I will spend a liitle time looking for some quick wins, but I will only submit any further changes if I think it is a clear win.

stephenworsley commented on lib/iris/experimental/ugrid/mesh.py:
```diff
@@ -975,77 +925,129 @@ def __ne__(self, other):
+            line("optional connectivities", 1)
+            for name, conn in optional_conns.items():
+                conn_string = conn.summary(shorten=True, linewidth=0)
+                line(f"{name}: {conn_string}", 2)
```
Could there be a case for grouping these optional connectivities together by their location/source element. e.g.

edge:
`edge_node_connectivity`
(optionally)
`edge_face_connectivity`

face:
`face_node_connectivity`
(optionally)
`face_edge_connectivity`
`face_face_connectivity`

with `boundary_node_connectivity` as a special case on its own.
In this way, connectivities would be grouped with other connectivities which share their dimensions.
(As an aside, I don't believe we currently keep hold of the NetCDF dimension associated with `boundary_node_connectivity`, I'm not sure if this is something we should be concerned about. I had a look at what UGRID says about them and it seems to suggest that `boundary_node_connectivity` can have ancillary variables attached which I don't know if we support. This is probably not core functionality, but it might be worth recording as an issue for later.)
With that said, I could also imagine an argument that something like `edge_face_connectivity` is just as much assiciated with faces as it is edges. I'm not necessarily asking for a change, but I thought it would be worth at least raising the point for discussion.

pp-mo replied:
At present I am purposely listing the _non_-optional 'edge_node' and 'face_node' connectivities in their "element section", as you suggest.

In my view, the "essential" connectivities really have a different scope to the 'optional' ones, and should be treated differently.  
I think If we put the 'optional' ones into their relevant sections, that's maybe not so clear.
In that case I think I would to list the required information first, and add the optional connectivities after the coords (if any).  Whether they need to be labelled 'optional' I'm not quite so sure.

As examples from an actual file ...
CURRENT
```
Mesh : 'Topology data of 2D unstructured mesh'
    topology_dimension: 2
    node
        node_dimension: 'nMesh2d_half_levels_node'
        node coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55298,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55298,)>
    edge
        edge_dimension: 'nMesh2d_half_levels_edge'
        edge_node_connectivity: <Connectivity: Maps every edge/link to two nodes that it connects. / (unknown)  <lazy>  shape(110592, 2)>
        edge coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(110592,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(110592,)>
    face
        face_dimension: 'nMesh2d_half_levels_face'
        face_node_connectivity: <Connectivity: Maps every face to its corner nodes. / (unknown)  <lazy>  shape(55296, 4)>
        face coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55296,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55296,)>
    optional connectivities
        face_face_connectivity: <Connectivity: Indicates which other faces neighbor each face / (no_unit)  <lazy>  shape(55296, 4)>
        face_edge_connectivity: <Connectivity: Maps every face to its edges. / (unknown)  <lazy>  shape(55296, 4)>
        edge_face_connectivity: <Connectivity: neighbor faces for edges / (unknown)  <lazy>  shape(110592, 2)>
    long_name: 'Topology data of 2D unstructured mesh'
    var_name: 'Mesh2d_half_levels'
```

OPTIONALS IN OWN SECTIONS
```
Mesh : 'Topology data of 2D unstructured mesh'
    topology_dimension: 2
    node
        node_dimension: 'nMesh2d_half_levels_node'
        node coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55298,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55298,)>
    edge
        edge_dimension: 'nMesh2d_half_levels_edge'
        edge_node_connectivity: <Connectivity: Maps every edge/link to two nodes that it connects. / (unknown)  <lazy>  shape(110592, 2)>
        edge coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(110592,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(110592,)>
        edge_face_connectivity: <Connectivity: neighbor faces for edges / (unknown)  <lazy>  shape(110592, 2)>
    face
        face_dimension: 'nMesh2d_half_levels_face'
        face_node_connectivity: <Connectivity: Maps every face to its corner nodes. / (unknown)  <lazy>  shape(55296, 4)>
        face coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55296,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55296,)>
        face_face_connectivity: <Connectivity: Indicates which other faces neighbor each face / (no_unit)  <lazy>  shape(55296, 4)>
        face_edge_connectivity: <Connectivity: Maps every face to its edges. / (unknown)  <lazy>  shape(55296, 4)>
    long_name: 'Topology data of 2D unstructured mesh'
    var_name: 'Mesh2d_half_levels'
```

OPTIONALS LABELLED AS SUCH
```
Mesh : 'Topology data of 2D unstructured mesh'
    topology_dimension: 2
    node
        node_dimension: 'nMesh2d_half_levels_node'
        node coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55298,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55298,)>
    edge
        edge_dimension: 'nMesh2d_half_levels_edge'
        edge_node_connectivity: <Connectivity: Maps every edge/link to two nodes that it connects. / (unknown)  <lazy>  shape(110592, 2)>
        edge coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(110592,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(110592,)>
        optional connectivities:
            edge_face_connectivity: <Connectivity: neighbor faces for edges / (unknown)  <lazy>  shape(110592, 2)>
    face
        face_dimension: 'nMesh2d_half_levels_face'
        face_node_connectivity: <Connectivity: Maps every face to its corner nodes. / (unknown)  <lazy>  shape(55296, 4)>
        face coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55296,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55296,)>
        optional connectivities:
            face_face_connectivity: <Connectivity: Indicates which other faces neighbor each face / (no_unit)  <lazy>  shape(55296, 4)>
            face_edge_connectivity: <Connectivity: Maps every face to its edges. / (unknown)  <lazy>  shape(55296, 4)>
    long_name: 'Topology data of 2D unstructured mesh'
    var_name: 'Mesh2d_half_levels'
```

WITH BOUNDARIES
```
Mesh : 'Topology data of 2D unstructured mesh'
    topology_dimension: 2
    node
        node_dimension: 'nMesh2d_half_levels_node'
        node coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55298,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55298,)>
    edge
        edge_dimension: 'nMesh2d_half_levels_edge'
        edge_node_connectivity: <Connectivity: Maps every edge/link to two nodes that it connects. / (unknown)  <lazy>  shape(110592, 2)>
        edge coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(110592,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(110592,)>
        optional connectivities:
            edge_face_connectivity: <Connectivity: neighbor faces for edges / (unknown)  <lazy>  shape(110592, 2)>
    face
        face_dimension: 'nMesh2d_half_levels_face'
        face_node_connectivity: <Connectivity: Maps every face to its corner nodes. / (unknown)  <lazy>  shape(55296, 4)>
        face coordinates
            <AuxCoord: longitude / (degrees)  <lazy>  shape(55296,)>
            <AuxCoord: latitude / (degrees)  <lazy>  shape(55296,)>
        optional connectivities:
            face_face_connectivity: <Connectivity: Indicates which other faces neighbor each face / (no_unit)  <lazy>  shape(55296, 4)>
            face_edge_connectivity: <Connectivity: Maps every face to its edges. / (unknown)  <lazy>  shape(55296, 4)>
    boundaries
        boundary_dimension: 'nMesh2d_half_levels_boundary'
        boundary_node_connectivity: <Connectivity: unknown / (unknown)  <lazy>  shape(1417, 2)>
    long_name: 'Topology data of 2D unstructured mesh'
    var_name: 'Mesh2d_half_levels'
```

pp-mo replied:
(After all that wordage)
I conclude that its perfectly possible, and it might make sense to put boundaries in their own section, and report the 'boundary_dimension', like this, but in that case I think we should support it in the Mesh object too.

But i think there are some CONs too : 
  * the existing output is more concise 
  * the existing output is possibly slightly clearer
  * I'm not convinced that we really know what the "status" of boundaries is : it seems to be a feature that is not fully developed + it's not clear quite how it is intended to be used or what for

lbdreyer replied:
I agree that this is definitely a worthwhile thing to discuss. The hard part is knowing what the best option is! I'm strugglling to come to a conclusion on this. I'm not terribly familiar with the area myself so finding it hard to decide what a user would expect. However, I am leaning towards keeping things as they are , primarily because I like the conciseness of the original output, also @stephenworsley's point:
> With that said, I could also imagine an argument that something like edge_face_connectivity is just as much assiciated with faces as it is edges. 
This is an interesting point! 

I think it might be simpler to keep the optional connectivities in their own section for now. But readdress this point at a later stage. Would you be okay with that @stephenworsley ?

stephenworsley replied:
Yeah, I think either way is workable and keeping it as it is currently is probably for the best for the moment. I imagine this could be something we could return to when we have a better idea of what to do with the boundaries.

lbdreyer reviewed (approved):
Great! I'm finding the code in the summary function very clear and easy to follow what's going on so I think this is a big improvement!

All my comments have been addressed so as far as I am concerned this is good to go in.
